dexlib2/src/main/java/org/jf/dexlib2/util/Preconditions.java

```diff
@@ -188,6 +195,48 @@ public static int checkVerificationError(int verificationError) {
         return verificationError;
     }
 
+    public static ImmutableList<? extends ImmutableSwitchElement> checkSequentialKeys(ImmutableList<? extends ImmutableSwitchElement> elements) {
+        TreeSet<Integer> sortedKeys = elements.stream()
```

Comment:
Instead of creating a sorted copy, why not just iterate through the items and check each one against the previous?

Reply:
> why not just iterate through the items and check each one against the previous

This would require that the provided SwitchElement list is already sorted by the keys. I think this is not guaranteed.
If the list is sorted and the keys are sequential, we can iterate through the list. If we see any number 'n', then the previously iterated value must be 'n-1' or there is one number missing and this is therefore not sequential key set.

If you don't have that order you must also look through the whole list and search for a one smaller element. This would be O(n²). My implementation should be O(n log(n)).

```java
public static ImmutableList<? extends ImmutableSwitchElement> checkSequentialKeys(ImmutableList<? extends ImmutableSwitchElement> elements) {
    int minKey = Integer.MAX_VALUE;
    for (ImmutableSwitchElement element : elements) {
        minKey = Math.min(element.getKey(), minKey);
    }

    for (ImmutableSwitchElement element : elements) {
        int key = element.getKey();
        if (key == minKey) continue;

        boolean existsPredecessor = false;
        for (ImmutableSwitchElement e : elements) {
            if (key == e.getKey() + 1) {
                existsPredecessor = true;
                break;
            }
        }

        if (!existsPredecessor) {
            throw new IllegalArgumentException("Keys are not sequential");
        }
    }

    return elements;
}
```

Reply:
It does currently require that they be sorted, although it's not documented as such, which should probably be improved.

For example, see here https://github.com/JesusFreke/smali/blob/2db1dec70385eedddd613ea7b426b8516973b5f9/dexlib2/src/main/java/org/jf/dexlib2/writer/InstructionWriter.java#L526. If the switch elements are unordered, this would produce incorrect output.

The way I see it, there are 2 options here - we can accept potentially unordered input, or require it to be ordered. Option 1 is the current state, and it also allows for a more efficiency when the caller already has the input in a sorted state. Given the nature of the input, this is likely to be the case. The main downside is that it makes the api slightly more onerous to use, in the uncommon case that the caller doesn't already have them sorted.

The other option makes the api a bit more friendly, but at the cost of making and sorting a copy of the input every time.

I think option 1 is acceptable here. It would also be good to document that the PackedSwitchPayload interface requires that the return value of getSwitchElements be sequential, and sorted by key.

Reply:
> It does currently require that they be sorted, although it's not documented as such, which should probably be improved.

That's true, I've once again looked through the classes and I've still found no hint to that requirement.



> Instead of creating a sorted copy, why not just iterate through the items and check each one against the previous?

I've now implemented that instead.